### PR TITLE
[2 of N] Changes for Making first render cycle faster and fixing test cases

### PR DESCRIPTION
**Pull request by ankityadav4**
## New Behavior
Improving first render cycle of cartesian chart 

Below are the improvements  this PR using CRA based approach:**

image
image

## Related Issue(s)

- Fixes #

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/871192b67e1bc8a68da1b4c55b4e0ee2aed0b604'>871192b67e1bc8a68da1b4c55b4e0ee2aed0b604</a></sub>
<div id="bundleSizeComment9423"/>

File: change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
Comment (AtishayMsft) [inline change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json:2]: Can you add the performance gain achieved due to this change.

File: change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "Improving first render cycle of cartesian chart",
+  "packageName": "@fluentui/react-charting",
+  "email": "ankityadav@microsoft.com",
+  "dependentChangeType": "patch"
+}
```
Comment (AtishayMsft) [inline change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json:7]: Can you help me understand this change.

File: change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (AtishayMsft) [inline change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json:1]: Lets update the PR with description and performance numbers.

File: packages/react-charting/src/components/AreaChart/AreaChart.types.ts [hunk 4]
```diff
@@ -53,6 +53,12 @@ export interface IAreaChartProps extends ICartesianChartProps {
    * Optimize area chart rendering for large data set.
    */
   optimizeLargeData?: boolean;
+
+  /**
+   * @default false
+   * The prop used to enable the perf optimization
+   */
+  enablePerfOptimization?: boolean;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/AreaChart/AreaChart.types.ts:61]: This change should go away if you merge master.

Comment (ankityadav4) [reply packages/react-charting/src/components/AreaChart/AreaChart.types.ts]: Done

File: change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json (hunk 3 above)
Comment (ankityadav4) [reply change/@fluentui-react-charting-bda51650-008f-417c-ab0d-62e7dd4a22fd.json:1]: Done

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 5]
```diff
@@ -738,7 +744,10 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
       labelWidth += Math.ceil(longestLabelWidth);
     }
 
-    let minChartWidth = this.margins.left! + this.margins.right! + labelWidth * (this._tickValues.length - 1);
+    let minChartWidth =
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:747]: Could you explain what this issue is

File: packages/react-examples/src/react-charting/LineChart/LineChart.Basic.Example.tsx [hunk 6]
```diff
@@ -186,6 +186,7 @@ export class LineChartBasicExample extends React.Component<{}, ILineChartBasicSt
             margins={margins}
             xAxisTickCount={10}
             allowMultipleShapesForPoints={this.state.allowMultipleShapes}
+            enablePerfOptimization={true}
```
Comment (AtishayMsft) [inline packages/react-examples/src/react-charting/LineChart/LineChart.Basic.Example.tsx:189]: Did you give the PR for testing after enabling this flag for all chart types.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 5 above)
Comment (ankityadav4) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:747]: The issue was due to the changes of PR: https://github.com/microsoft/fluentui/pull/28239. Changes done in the above PR are dependent on the first render cycle computations. So after merging it with my changes for optimization of first render cycle the chart was not rendering due to NPE issues at various places.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts [hunk 7]
```diff
@@ -568,4 +568,10 @@ export interface IModifiedCartesianChartProps extends ICartesianChartProps {
    *@default false
    *Used for showing complete y axis lables   */
   showYAxisLables?: boolean;
+
+  /**
+   * @default false
+   * Used to test the Performance optimization code befor merging it to master.
```
[the remark this answers is not in the document]
Comment (ankityadav4) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts]: Done

File: packages/react-examples/src/react-charting/LineChart/LineChart.Basic.Example.tsx (hunk 6 above)
Comment (ankityadav4) [reply packages/react-examples/src/react-charting/LineChart/LineChart.Basic.Example.tsx:189]: No, earlier i have tested it locally for all the variants.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 5 above)
Comment (AtishayMsft) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:747]: Got it.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 8]
```diff
@@ -713,7 +719,7 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
       labelWidth += Math.ceil(longestLabelWidth * Math.cos(Math.PI / 4));
     }
     // Case: truncated labels
-    else if (this.props.showXAxisLablesTooltip) {
+    else if (this.props.showXAxisLablesTooltip && this._tickValues) {
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:722]: Shouldn't this whole function be get called in the second render cycle only. I see this used in rotated labels case also.

Comment (ankityadav4) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx]: Done

**Review by AtishayMsft — APPROVED** [review]

**Pull request merged**